### ISSUE TITLE
availWidth, availHeight Not Reliable on Mac OS X

### ISSUE THREAD

**Issue by stuart-lam** [Valve/fingerprintjs2#99]
Hey man, thank you for this kick ass project! I found this availWidth/Height is not so reliable on Mac OS because I accidently see my fingerprint changed while in the same browser, not even change anything and not even refresh. I investigated the issue and found that when you launch an App in "Lauchpad" then switch to "Dashboard" in Mac OS X, switch back then come back to the browser, Click the button "Get My Fingerprint" => Fingerprint changed.

The real issue is this:
{"key":"available_resolution","value":[1440,812]}
Changed to:
{"key":"available_resolution","value":[1440,810]}

Why should we use it to calculate fingerprint?

Cheers,
Stuart

**Comment by stuart-lam** [comment]
Checkout my screenshot:

<img width="830" alt="screen shot 2015-12-13 at 7 03 53 pm" src="https://cloud.githubusercontent.com/assets/16276221/11766955/f6918b9e-a1ce-11e5-94cf-2a39ddc69ca9.png">
<img width="896" alt="screen shot 2015-12-13 at 7 04 41 pm" src="https://cloud.githubusercontent.com/assets/16276221/11766956/f69413aa-a1ce-11e5-9d9d-c1a69d84de81.png">

**Comment by Valve** [comment]
Hi, 

thanks for using the library!

First of all, which browser and version were you using to do this?

We're using the availHeight/Width to not only be able to get the screen resolution but to get the available resolution, which can differ because of the taskbar/dock location or various other settings, such as font size (which increases the window title). So this option is valuable. As for your bug, I'll look into it when you let me know your browser and version.

**Comment by stuart-lam** [comment]
Hi Valve! Thank you a lot for your efforts into this! But I don't think this is a bug in your script but a bug in Mac OS, switching between Launchpad / Dashboard / Desktop(s) in Mac OS return in availWidth/Height changes. Anyway man, this is my info:

```

Google Chrome	47.0.2526.80 (Official Build) (64-bit)
Revision	9fd42dd40d33146ff9afd92f51714f0426c1247c-refs/branch-heads/2526@{#505}
OS	Mac OS X 
Blink	537.36 (@9fd42dd40d33146ff9afd92f51714f0426c1247c)
JavaScript	V8 4.7.80.25
Flash	20.0.0.228
User Agent	Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.80 Safari/537.36
```

**Comment by Valve** [comment]
I cannot reproduce it, always getting consistent fingerprints. Which OSX version are you on? 
I'm using 10.10.5 + Google Chrome 47

**Comment by Valve** [comment]
I cannot reproduce it, probably I haven't upgraded to El Capitan yet. Working fine for me on 10.10.5.

I'll leave this issue open to look into it after I upgrade

**Comment by stuart-lam** [comment]
yeah man, I'm in El Capitan 10.11.2 now and it's still getting different Fingerprints. You have to do 100% like what I did (in orders) to reproduce the problem:
```
1. Go to http://valve.github.io/fingerprintjs2/ with Chrome, get you fingerprint first.
2. Open your Launchpad > Other > Activity Monitor
3. Open your Launchpad > Other > Dashboard
4. Switch back to Chrome.
5. Click the "Get my fingerprint" button.
```
You will see the different fingerprint!

**Comment by Valve** [comment]
yep, just repeated your steps, FP is consistent. Will wait for the El Capitan upgrade to reproduce it.

**Comment by nuschk** [comment]
Guys, I was also able to reproduce this issue. I'm using Chrome 47.0.2526.111 on MacOS El Capitan (10.11.2). 

For me, it happens when I open an incognito window, check the fingerprint, then quit Chrome and repeat.

The behavior matches what @stuart-lam describes: the available resolutions are just a few pixels of for each try.

I also see changing plugin orderings, so I've already added sortPluginsFor: [/.*/] in my config. So, for me, it would be cool to be able to disable available resolutions checking through the options as well. I might add that, for our application, I'm more than willing to trade a few false positives for less false negatives. Fine grained control over what gets checked and what doesn't really helps with tweaking fingerprint in this regard.

**Comment by Valve** [comment]
@nuschk thanks for you observations. Currently I'm in the process of relocating to another country so I can't really tell when I will be able to return to the lib. Probably only in a month or two

**Comment by nuschk** [comment]
Yeah, no worries, I can always fork. For your reference, I've added a pull request (#115) which helps me to just disable this key separately from basic resolution checking. Works perfectly for me.
